### PR TITLE
feat: ccr sftp file download

### PR DESCRIPTION
**Pull request by tomshag**
Har opprettet prosjekt som tar seg av tilkobling til SFTP server og gjør nedlasting av nye filer basert på provided RunId.

## Description
Koden består av ErDataTransClient som opprettes med Host, Usr, Pw og remotepath for å sette opp tilkobling,
"GetNewFiles" kjøres for å hente nye filer.
Dersom LastRunId gis med henter den kun nye filer som stemmer overens med forventet Id.
Filer blir automatisk renamet med "baj07111Downloaded.txt" når de er blitt hentet. (For å skille mellom Altinn 2 og Altinn 3, siden Altinn 2 renamer til baj07111.txtretrieved)

Hvis en fil kalt "baj07111.txtretrieved" blir hentet betyr det at den har blitt hentet av Altinn 2, men den vil også bli behandlet i Altinn 3 og bli renamet til "baj07111Downloaded.txtretrieved".

Har også laget en "GetSpecificFile" for å kunne hente ned en spesifikk fil om det er nødvendig.

Hva koden ikke gjør:
Koden gjør ikke Initialisering av nedlasting eller kall for å viderebehandle nedlastet data, den returnerer kun filstrømmer for nye filer med tilhørende filnavn i en tuple.
Koden lagrer ikke sekvens nr eller fildata.

## Related Issue(s)
- #798 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [x] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

## Summary by CodeRabbit

* **New Features**
  * Added ER DataTrans SFTP integration to retrieve and manage register data files from remote servers.

* **Tests**
  * Added unit tests covering file discovery, retrieval, renaming and edge-case filename parsing.

* **Chores**
  * Solution and build configurations updated to include the new integration project and platform mappings.
  * Added SSH.NET SFTP client dependency to support the integration.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Altinn/altinn-register/pull/865)

### CODE REVIEW

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+            int runId = ErDataTransClient.GetRunIdFromFileName(file.Name);
+            if (!file.FullName.Contains("Downloaded")
+                && (lastRunId == -1 || runId == lastProcessedRunId + 1))
+            {
+                Stream stream = new MemoryStream();
+                _client.DownloadFile(file.FullName, stream);
+                stream.Position = 0; // Reset stream position after download
+                streams.Add(new Tuple<string, Stream>(file.Name, stream));
+                _client.RenameFile(file.FullName, file.FullName.Replace(".txt", "Downloaded.txt")); // Mark file as downloaded
```
[the remark this answers is not in the document]
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: baj05778.txtretrieved should be turned into baj05778Downloaded.txtretrieved since the file is retrieved by 2 different systems that should be able to function independently.

File: Altinn.Register.sln [hunk 2]
```diff
@@ -355,6 +369,7 @@ Global
 		{9CDE5E81-EBC7-458C-A97E-F55F8BF4A4F1} = {44C13623-0995-2183-17AD-382970CA8A29}
 		{2E11F4DE-5E45-416E-BE04-222DDEA21A67} = {9F856603-52D5-E1B8-657B-994E194BE88D}
 		{410B52EF-62C5-4506-B6D3-802F85BDDA92} = {9F856603-52D5-E1B8-657B-994E194BE88D}
+		{AE58849C-EBDE-4FDC-8830-0402EAE592C7} = {77D98184-A1B7-9DC0-C6B7-1D19D6300698}
```
[the remark this answers is not in the document]
Comment (tomshag) [reply Altinn.Register.sln]: This should now be fixed

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        // Extract runId from filename, e.g., "baj05778.txt" -> 5778
+        int bajIndex = filename.IndexOf("baj");
+        int dotIndex = filename.IndexOf("Downloaded.txt");
+        if(dotIndex == -1)
+        {
+            dotIndex = filename.IndexOf(".txt");
+        }
+
+        string runIdPart = filename[(bajIndex + 3)..dotIndex]; // Assuming filename format is always "bajXXXXX.txt"
```
[the remark this answers is not in the document]
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: this should now be fixed

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            foreach (var file in files.OrderBy(f => f.Name))
+            {
+                if (file.Name.Contains("Downloaded"))
+                {
+                    continue;
+                }
+
+                int runId = ErDataTransClient.GetRunIdFromFileName(file.Name);
+                if (!file.FullName.Contains("Downloaded")
+                    && (lastRunId == -1 || runId == lastProcessedRunId + 1))
```
[the remark this answers is not in the document]
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: For now unexpected filenames should stop processing

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                {
+                    Stream stream = new MemoryStream();
+                    _client.DownloadFile(file.FullName, stream);
+                    stream.Position = 0; // Reset stream position after download
+                    streams.Add(new Tuple<string, Stream>(file.Name, stream));
+                    _client.RenameFile(file.FullName, file.FullName.Replace(".txt", "Downloaded.txt")); // Mark file as downloaded
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Burde vi omdøpe fila før vi har "committa" den på vår side? Hva om server kræsjer? Eller kubernetes tar liv av register for vi har for store `MemoryStream`s 😛

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                if (file.Name.Contains("Downloaded", StringComparison.InvariantCultureIgnoreCase) || !file.Name.StartsWith("baj", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    continue;
+                }
+
+                int runId = ErDataTransClient.GetRunIdFromFileName(file.Name);
+                if (!file.FullName.Contains("Downloaded")
+                    && (lastRunId == -1 || runId == lastProcessedRunId + 1))
+                {
+                    Stream stream = new MemoryStream();
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Disse filene er vel potensiellt store? Er det ikke dumt å ha den (potensiellt flere av dem) in-memory?

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        if (!_client.IsConnected)
+        {
+            _client.Connect();
+        }
+
+        try
+        {
+            string fullPath = $"{_remotePath}/{filename}";
+            Stream stream = new MemoryStream();
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Samme komentarer som over gjelder.

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        _client = client;
+        _remotePath = remotePath;
+    }
+
+    /// <summary>
+    /// Retrieves new files from the SFTP server based on the last processed runId. It checks for files in the specified remote path, and if a file's runId is exactly one greater than the last processed runId, it downloads the file and marks it as downloaded by renaming it. The method returns a list of tuples containing the filename and its content as a stream.
+    /// </summary>
+    /// <param name="lastRunId">Last RunId</param>
+    /// <returns>List of tuples containing the filename and its content as a stream.</returns>
+    public List<Tuple<string, Stream>> GetNewFiles(int lastRunId = -1)
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Alle metoder som interagerer med netverk må være async.

Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Jeg tror å snu på kontrakten her er fornuftig. I stedet for at du får tilbake 1 (eller flere) streams, gjør denne til:

```csharp
/// <returns><see langword="true"/> if there are more files to process, otherwise <see langword="false"/></returns>
public async Task<bool> ProcessNextFile(PipeWriter writer, int lastRunId = -1)
```

Da kan fila flyttes *etter* at vi er ferdig med den. Og metoden kan kalles på nytt hvis den så lenge den returnerer `true`.

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                    Stream stream = writer.AsStream();
+                    await _client.DownloadFileAsync(file.FullName, stream, cancellationToken); // Download file content to the provided stream
+                    await writer.FlushAsync(cancellationToken); // Ensure all data is flushed before disconnecting
+                    await writer.CompleteAsync(); // Complete the writer to signal no more data will be written
+                    return true;
+                }
+            }
+
+            return false;
+        }
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs:84]:
```csharp
catch (Exception e)
{
    await writer.CompleteAsync(e);
    throw;
}
```

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                {
+                    Stream stream = writer.AsStream();
+                    await _client.DownloadFileAsync(file.FullName, stream, cancellationToken); // Download file content to the provided stream
+                    await writer.FlushAsync(cancellationToken); // Ensure all data is flushed before disconnecting
+                    await writer.CompleteAsync(); // Complete the writer to signal no more data will be written
+                    return true;
+                }
+            }
+
+            return false;
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs:83]: The should probably be closed here as well.

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    public async Task<bool> GetNextFileAsync(PipeWriter writer, int lastRunId, CancellationToken cancellationToken = default)
+    {
+        if (!_client.IsConnected)
+        {
+            _client.Connect();
+        }
+
+        try
+        {
+            IEnumerable<ISftpFile> files = _client.ListDirectory(_remotePath).Where(f => !f.IsDirectory && !f.IsSymbolicLink);
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs]: ListDirectoryAsync - we're dealing with network drives here, and these are nerwork operations.

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    /// </summary>
+    /// <param name="writer">PipeWriter to write the file content.</param>
+    /// <param name="lastRunId">Last RunId</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>Tuple containing the filename and its content as a stream.</returns>
+    public async Task<bool> GetNextFileAsync(PipeWriter writer, int lastRunId, CancellationToken cancellationToken = default)
+    {
+        if (!_client.IsConnected)
+        {
+            _client.Connect();
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs]: ConnectAsync

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+                    await writer.CompleteAsync(); // Complete the writer to signal no more data will be written
+                    return true;
+                }
+            }
+
+            return false;
+        }
+        finally
+        {
+            _client.Disconnect();
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs:92]: DisconnectAsync

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        }
+    }
+
+    /// <summary>
+    /// Marks a file as downloaded by renaming it on the SFTP server. It connects to the SFTP server, checks for files in the specified remote path, and if a file's runId is exactly one greater than the last processed runId, it renames the file by replacing ".txt" with "Downloaded.txt". After processing, it disconnects from the SFTP server. The method returns true if a file was marked as downloaded, otherwise false.
+    /// </summary>
+    /// <param name="lastRun">The last processed runId. Used to identify which file should be renamed.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>A task representing the asynchronous operation.</returns>
+    public async Task<bool> MarkFileAsDownloadedAsync(int lastRun, CancellationToken cancellationToken = default)
```
Comment (Alxandr) [inline src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs]: Basically the same as the method above.

File: src/apps/Altinn.Register/test/Altinn.Register.Tests/UnitTests/CcrDataTransferTests.cs [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+        var result = await client.GetNextFileAsync(writer, -1, TestContext.Current.CancellationToken);
+
+        // Assert
+        Assert.True(result);        
+        var readResult = await reader.ReadAsync(TestContext.Current.CancellationToken);
+        var buffer = readResult.Buffer;
+        string resultContent = Encoding.ASCII.GetString(buffer.ToArray());
+        Assert.Equal(text, resultContent);
+
+        // Verify rename was called
```
Comment (Alxandr) [inline src/apps/Altinn.Register/test/Altinn.Register.Tests/UnitTests/CcrDataTransferTests.cs]: Verifies that rename was *not* called 😛 - I got a bit confused here

**Review by Alxandr — APPROVED** [review]

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs (hunk 5 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Dette er nå fikset og renaming av fil er et eget kall

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs (hunk 6 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Leverer fil over Pipewriter

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs (hunk 8 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Gjorde metdoene asynce

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs (hunk 7 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.ErDataTrans/ErDataTransClient.cs]: Leverer fil over Pipewriter

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs (hunk 9 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs:84]: utført

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs (hunk 11 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs]: resolved

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs (hunk 13 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs:92]: there is no disconnectasync, other asyncs have been resolved

File: src/apps/Altinn.Register/test/Altinn.Register.Tests/UnitTests/CcrDataTransferTests.cs (hunk 15 above)
Comment (tomshag) [reply src/apps/Altinn.Register/test/Altinn.Register.Tests/UnitTests/CcrDataTransferTests.cs]: updated comment

File: src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs (hunk 14 above)
Comment (tomshag) [reply src/apps/Altinn.Register/src/Altinn.Register.Integrations.Ccr.FileImport/CcrDataTransfer.cs]: resolved
